File: content/cgi-bin/cgi.php
#!/usr/bin/env php
<?php

$data = '';
$status = 200;
$methode = $_SERVER['REQUEST_METHOD'];
if (!strlen($methode))
    exit (1);
$dir = $_SERVER['UPLOAD_DIR'];
if (!strlen($dir))
    // exit (1);
    $dir = "/Users/ykhayri/Desktop/wsl_webserv/content/cgi-bin/uploads/";


if ($methode === 'POST') {
    while (($line = fgets(STDIN)) !== false) {
        $data .= $line;
        $data .= "\n";
    }
}
else {
    $data = $_SERVER['QUERY_STRING'];
    if (!strlen($data))
        exit (1);
}


$data = explode('&', $data);
$disp = "";
foreach ($data as $key=>&$x) {
    $x = str_replace("=", ": ", urldecode($x));
    if ($methode == 'DELETE') {
        $tmp = explode(": ", $x);
        if ($tmp[0] == 'fileName') {
            $rem = $dir . $tmp[1];
            if (!file_exists($rem)){
            // fwrite(STDERR, "====================================> This is an error message\n" . $rem);
                exit(1);
            }
        }
        else
            exit(1);
    }
    if ($key > 1)
        $disp .= "<h2>{$x}</h2>";
}
if ($methode == 'DELETE') {
    foreach ($data as $key=>&$x) {
        $x = str_replace("=", ": ", urldecode($x));
        $tmp = explode(": ", $x);
        if ($tmp[0] == 'fileName') {
            $rem = $dir . $tmp[1];
            if (!unlink($rem))
                exit(1);
            $status = 204;
        }
    }
}

$disp .= "<h2>{$s}</h2>";
$body = <<<EOD
    <!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>CGI</title>
        <link rel="stylesheet" href="/cgi-bin/forms/styles/style.css" />
    </head>
    <body>
        <div class="response">
            <h1>Methode: {$methode} </h1>
            {$disp}
        </div>
    </body>
    </html>
EOD;
if ($status == 204)
    $body = '';
echo "HTTP/1.1 " . $status . " OK\r\nContent-Length: " . strlen($body) . "\r\nContent-Type: text/html\r\n\r\n" . $body;

?>

File: content/cgi-bin/cgi.py
#!/usr/bin/env python3

import os
import urllib.parse
import sys
import re

if "REQUEST_METHOD" in os.environ:
	meth = os.environ['REQUEST_METHOD']
else:
	exit (1)

if "UPLOAD_DIR" in os.environ:
	dir = os.environ['UPLOAD_DIR']
else:
	exit (1)

data = ''
disp = ''
status = ["200", "OK"]

def addH2():
	tmp = ''
	for index, elem in enumerate(data):
		if (index > 1 and len(elem) > elem.find('=') + 1):
			tmp = tmp + f'<h2>{urllib.parse.unquote(elem).replace("+", " ").replace("=", ": ", 1)}</h2>'
	return tmp

if (meth == 'POST'):
	data = sys.stdin.buffer.read()
	data = data.decode("utf8").split('&')
else:
	if "QUERY_STRING" in os.environ:
		data = os.environ['QUERY_STRING']
		data = data.split('&')
	else:
		exit (1)
	if (meth == "DELETE"):
		myfl = ''
		for index, elem in enumerate(data):
			tmp = elem.split("=")
			if (tmp[0] == "fileName"):
				myfl = dir + tmp[1]
				if(not os.path.exists(myfl)):
					exit(1)
			else:
				exit(1)
		
		for index, elem in enumerate(data):
			tmp = elem.split("=")[1]
			try:
				myfl = dir + tmp
				os.remove(myfl)
				status = ["204", "No Content"]
			except Exception:
				exit(1)



disp = " ".join(status)
if (meth != "DELETE"):
	disp = addH2()

body = f"""
	<!DOCTYPE html>
	<html lang="en">

	<head>
		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		<title>CGI</title>
		<link rel="stylesheet" href="/cgi-bin/forms/styles/style.css" />
	</head>

	<body>
		<div class="response">
			<h1>Methode: {meth}</h1>
			{disp}
		</div>
	</body>

	</html>
"""

print(f"HTTP/1.1 {' '.join(status)}\r\nContent-Length: {len(body)}\r\nContent-Type: text/html\r\n\r\n{body}", end="")
